Error code

justinjung04 opened the pull request:
<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

### Description

Currently the API always returns http code 422 for engine execution error, regardless of the HTTP error codes embedded in the gRPC error chain.

This PR improves it to retrieve the first valid HTTP error code, if applicable, and return it.

Before the PR:

```
Body: {"status":"error","data":"test","errorType":"execution","error":"some error: rpc error: code = Code(429) desc = message"}
Status Code: 422

Body: {"status":"error","data":"test","errorType":"execution","error":"some error: rpc error: code = Code(998) desc = message"}
Status Code: 422
```

After the PR:

```
Body: {"status":"error","data":"test","errorType":"execution","error":"some error: rpc error: code = Code(429) desc = message"}
Status Code: 429

Body: {"status":"error","data":"test","errorType":"execution","error":"some error: rpc error: code = Code(998) desc = message"}
Status Code: 422 // still 422 since 998 is an invalid HTTP status
```

### Testing

Added unit tests

machine424 commented:
Thanks for this.
I don't think this is meant for Prometheus itself, which means this will always result in unnecessary traversals of the `errorExec` chains. 
For library use, we need to consider a different approach that doesn't penalize Prometheus.

justinjung04 commented:
Thank you @machine424 for taking a look.
To avoid the traversals, would you be more willing to have a struct that contains error code and error chain (and make it a standard way of returning error from downstream), such that Prometheus could just read the error code directly?

justinjung04 commented:
Added new struct

```
type ErrorWithStatusCode struct {
	StatusCode int
	Err        error
}
```

and downstream components (prometheus dependencies) can do the following to specify HTTP error code on engine execution error:

```
if errWithStatusCode, ok := grpcutil.ErrorWithHTTPStatusCode(code, err); ok {
  return errWithStatusCode
}
return err
```

machine424 commented:
Thanks, but I still believe that `util/grpcutil` doesn't belong in the Prometheus codebase.
Instead, we could allow customization of the `apiError -> status code` mapping, assuming other maintainers agree to open that door.

I hadn't shared my opinion earlier, but as an end user I’d be confused e.g. if an API occasionally returned a `429` status, because the API itself is rate limited by an underlying service. Could you share more about your use case?

justinjung04 commented:
I believe another option is to add a new field in the `API` struct that contains an error code override function, but only applicable to `errorExec` type. Please let me know what you would prefer!

machine424 commented:
Thanks for sharing the details.
It's about making tradeoffs with such changes, we want the code to be customizable with less overhead on your end, but we also want to keep the code on our end simple and readable.
we could try replacing the switch with a lookup from `map[errorType]func(errorType)int` for which we can allow dependents to override entries, but we should check that the overhead is acceptable, if not we'd need to go for a default `switch` that you can override as a whole...

justinjung04 commented:
- Created `ErrorTypeToStatusCode` which can be passed to `v1.NewAPI` to override the status code based on error type and the error itself.
- Exposed all `errorType`
- Added unit tests
- Added changelog

justinjung04 commented:
@machine424 Updates have been made and conflicts have been resolved. Could you take another look?

justinjung04 commented:
All comments addressed!

machine424 commented on web/api/v1/api_test.go:
```diff
@@ -4123,22 +4123,65 @@ func TestRespondSuccess_DefaultCodecCannotEncodeResponse(t *testing.T) {
-		api.respondError(w, &apiError{errorTimeout, errors.New("message")}, "test")
-	}))
-	defer s.Close()
+	type test struct {
```
thanks, I think we can simplify it a little bit
```diff
diff --git a/web/api/v1/api_test.go b/web/api/v1/api_test.go
index 18305ff3d..8e926c9e6 100644
--- a/web/api/v1/api_test.go
+++ b/web/api/v1/api_test.go
@@ -4125,48 +4125,37 @@ func TestRespondSuccess_DefaultCodecCannotEncodeResponse(t *testing.T) {
 func TestRespondError(t *testing.T) {
        type test struct {
                errType             errorType
-               err                 error
                resCode             int
-               msg                 string
                errTypeToStatusCode ErrorTypeToStatusCode
        }
 
        tests := map[string]test{
                "timeout should return 503 (ServiceUnavailable)": {
                        errType: ErrorTimeout,
-                       err:     errors.New("message"),
                        resCode: http.StatusServiceUnavailable,
-                       msg:     "message",
                },
                "execution error without override should return 422 (UnprocessableEntity)": {
                        errType: ErrorExec,
-                       err:     errors.New("message"),
                        resCode: http.StatusUnprocessableEntity,
-                       msg:     "message",
                },
                "errorTypeToStatusCode override works as expected": {
                        errType: ErrorExec,
-                       err:     errors.New("message"),
                        errTypeToStatusCode: map[errorType]func(error) int{
                                ErrorExec: func(_ error) int {
                                        return http.StatusTooManyRequests
                                },
                        },
                        resCode: http.StatusTooManyRequests,
-                       msg:     "message",
                },
        }
 
        for name, tc := range tests {
                t.Run(name, func(t *testing.T) {
                        s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
-                               api := API{}
-                               err := tc.err
-                               api.errorTypeToStatusCode = map[errorType]func(error) int{}
-                               for errType, handler := range tc.errTypeToStatusCode {
-                                       api.errorTypeToStatusCode[errType] = handler
+                               api := API{
+                                       errorTypeToStatusCode: tc.errTypeToStatusCode,
                                }
-                               api.respondError(w, &apiError{tc.errType, err}, "test")
+                               api.respondError(w, &apiError{tc.errType, errors.New("message")}, "test")
                        }))
                        defer s.Close()
 
@@ -4179,7 +4168,7 @@ func TestRespondError(t *testing.T) {
                        require.Equal(t, want, have, "Return code %d expected in error response but got %d", want, have)
                        h := resp.Header.Get("Content-Type")
                        require.Equal(t, "application/json", h, "Expected Content-Type %q but got %q", "application/json", h)
-                       require.JSONEq(t, fmt.Sprintf(`{"status": "error", "data": "test", "errorType": "%s", "error": "%s"}`, tc.errType, tc.msg), string(body))
+                       require.JSONEq(t, fmt.Sprintf(`{"status": "error", "data": "test", "errorType": "%s", "error": "message"}`, tc.errType), string(body))
                })
        }
 }
```

machine424 reviewed (approved):
lgtm, thanks again.
(I just have some suggestions regarding a test)

machine424 commented:
cc @bboreham as a "prometheus as a library" user, FYI

bboreham commented on web/api/v1/api.go:
```diff
@@ -76,17 +77,19 @@ const (
+	ErrorNotAcceptable errorType = "not_acceptable"
 )
 
+type ErrorTypeToStatusCode map[errorType]func(error) int
```
Is it valid to export this and leave `errorType` unexported?  How would someone implement it?

justinjung04 replied:
Although the type is not exposed, the enums themselves have been changed to be exposed. So you would use it like this:
```
api := API{
        ...
	errorTypeToStatusCode: ErrorTypeToStatusCode{
		ErrorExec: func(err error) int {
			if strings.Contains(err.Error(), "too many requests") {
				return 429
			}
			return 422
		},
	},
}
```

bboreham commented:
If you're going to change this, why go via strings at all?  Isn't it logically more like an enum?

justinjung04 commented:
Oh I see, I believe we still need the string consts since they're used for creating apiError:

```
type apiError struct {
	typ errorType
	err error
}

func (e *apiError) Error() string {
	return fmt.Sprintf("%s: %s", e.typ, e.err)
}
```

To avoid string hashing, I can create another set of enum (const with int type), and add that to the apiError struct like this?

```
// existing string consts
type errorTypeStr string
const (
	errorTimeoutStr       errorTypeStr = "timeout"
	errorCanceledStr      errorTypeStr = "canceled"
	...
)

// new enum exported
type errorType int
const (
	ErrorTimeout       errorType = iota
	ErrorCanceled
	...
)

// update apiError struct to include both error string and type
type apiError struct {
	typ errorType
	typStr errorTypeStr
	err error
}

...

if handler, ok := api.errorTypeToStatusCode[apiErr.typ]; ok {
	code = handler(apiErr.err)
} else {
	switch apiErr.typStr {
	case errorTimeoutStr:
        ...
	}
}
```

Please let me know if this alternative looks good to you

justinjung04 commented:
Updated `errorType` to a struct that contains both int and string. The int values for all error types are exported, and `ErrorTypeToStatusCode` uses the exported error type int as the map key.

justinjung04 commented:
@machine424 @bboreham Could you take another look?

bboreham reviewed (approved):
Thanks; I guess we can live with this but it still seems a little over-complicated.

Do we need both tests?  They seem to be doing about the same thing for the new functionality.

justinjung04 commented on web/api/v1/api.go:
```diff
@@ -73,20 +74,39 @@ const (
+	errorNotAcceptable = errorType{ErrorNotAcceptable, "not_acceptable"}
 )
 
+type ErrorTypeToStatusCode map[errorNum]func(error) int
```
That's a good point. The map was really to filter out overriden handlers, but with `func(errorNum error) (bool, int)` we can rely on the overrider to just return quickly for errors that they want to fall back. I'll make the change.

justinjung04 commented:
@machine424 @bboreham could I get final reviews?
